internal/agent/agent.go

```diff
@@ -274,24 +276,56 @@ func checkCollectorConnectivity(ctx context2.Context, cfg *config.Config, retrie
 	return
 }
 
-func getSampleMatcher(c *config.Config) func(interface{}) bool {
+func newSampleMatcher(c *config.Config, ffRetriever feature_flags.Retriever) func(interface{}) bool {
+	// configuration option always takes precedence over FF and matchers configuration
+	if c.EnableProcessMetrics != nil && *c.EnableProcessMetrics == false {
+		alog.Debug("EnableProcessMetrics is FALSE, process metrics will be DISABLED")
+		return func(sample interface{}) bool {
+			// no process samples will be sent to backend
+			return false
+		}
+	}
+
+	if c.EnableProcessMetrics != nil && *c.EnableProcessMetrics == true {
+		ec := sampler.NewMatcherChain(c.IncludeMetricsMatchers)
+		if ec.Enabled {
+			alog.Debug("EnableProcessMetrics is TRUE and rules ARE defined, process metrics will be ENABLED for matching processes")
+			return func(sample interface{}) bool {
+				return ec.Evaluate(sample)
+			}
+		} else {
+			alog.Debug("EnableProcessMetrics is TRUE and no rules defined, ALL process metrics will be ENABLED")
+			return func(sample interface{}) bool {
+				// ALL process samples will be sent to backend
+				return true
+			}
+		}
+	}
+
+	// if config option is not set, check if we have rules defined. those take precedence over the FF
 	ec := sampler.NewMatcherChain(c.IncludeMetricsMatchers)
 	if ec.Enabled {
+		alog.Debug("EnableProcessMetrics is TRUE and rules ARE defined, process metrics will be ENABLED for matching processes")
 		return func(sample interface{}) bool {
 			return ec.Evaluate(sample)
 		}
+	} else {
+		// configuration option is not defined, check feature flag
+		if enabled, exists := ffRetriever.GetFeatureFlag(handler.FlagFullProcess); exists {
```

Comment:
I believe the ffRetriever should be included inside the returned func. because the Command Channel is async, if this function is executed before the first command channel query we might end up not reading the proper FF value

Reply:
nice catch! ffRetriever was extracted without realising at https://github.com/newrelic/infrastructure-agent/pull/27/files#diff-93c6ab57eec49755878074307ef401e4R315

fixed at https://github.com/newrelic/infrastructure-agent/pull/26/commits/09f79f77d3e767474c24a73e884d6a94ba82e369